Prune stale core domain receipts

liuchengxu opened the pull request:
This PR is primarily for resolving a TODO which is to prune the stale state of core domain receipts that are deep enough on the system domain. Recommended to review commit by commit and have a look at the detailed commit message which helps understand the changes.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](../CONTRIBUTING.md)

liuchengxu committed:
Track primary block hash and prune core domain receipts in pallet-domain-registry

One notable thing is that each core domain receipt in pallet-domain-registry
can be ensured to point to a valid primary block which has been checked on the
primary chain, so the storage item `BlockHash` in pallet-domain-registry can be
updated based on the new receipt.

liuchengxu committed:
Fix pre_dispatch_submit_bundle() in pallet-domains

The problem is that pallet-domains only maintains the system receipts,
thereby only the system receipt head is available in pallet-domains. In
terms of the core domain receipts, it can only be verified whether the
receipt points to a valid primary block within pallet-domains.

liuchengxu committed:
Check core domain missing receipt in pre_dispatch_submit_core_bundle()

nazar-pc commented on crates/pallet-domains/src/lib.rs:
```diff
@@ -530,30 +530,40 @@ impl<T: Config> Pallet<T> {
-            // Non-best receipt
-            if receipt.primary_number <= best_number {
-                if BlockHash::<T>::get(receipt.primary_number) != receipt.primary_hash {
+        if signed_opaque_bundle.domain_id().is_system() {
```
I don't think I follow how this fixes anything. Commit description wasn't sufficient for me. I don't see why would other checks present here fail for non-system domain.

liuchengxu replied:
The check of validity of pair (receipt.primary_number, receipt.primary_hash) is the same for both system and core domain, which ensures the receipt points a primary block in the history. However, `ReceiptHead` in pallet-domain is for the system domain only and shouldn't be used to check the missing parent core domain receipt, the missing receipt of core domain is checked in pallet-domain-registry as their head receipt is maintained there.

nazar-pc replied:
I think I misread the code a bit. Without full scope this looks fine to me.

nazar-pc commented on domains/pallets/domain-registry/src/lib.rs:
```diff
@@ -179,6 +179,15 @@ mod pallet {
+    ///
+    /// NOTE: The oldest block hash will be pruned once the oldest receipt is pruned. However, if the
+    /// execution chain stalls, i.e., no receipts are included in the primary chain for a long time,
+    /// this mapping will grow indefinitely.
```
But this pallet is included in execution chain itself. Comment was copy-pasted and likely needs to be adjusted.

liuchengxu replied:
Updated docs and _fixed_ it since it must be domain-aware as well, otherwise in case some core domain stalls but the other domains still moves forward as normal, this state become corrupted.

liuchengxu committed:
Make BlockHash in pallet-domain-registry domain-aware

While updating the docs, I realized that `BlockHash` must be
domain-aware too otherwise if any core domain stalls but the other
domains keep moving steadily, then the state corrupts.

nazar-pc reviewed (approved):
Will merge once CI passes
